File: www/encode.js
import { Vec } from "./math.js";

class Encoder {
    _dv = new DataView(new ArrayBuffer(1024));
    _offset = 0;

    constructor() {
        this.reset();
    }

    reset() {
        this._offset = 0;
    }

    writeUint8(val) {
        this._dv.setUint8(this._offset, val);
        this._offset += 1;
    }

    writeInt32(val) {
        this._dv.setInt32(this._offset, val);
        this._offset += 4;
    }

    writeFloat64(val) {
        this._dv.setFloat64(this._offset, val);
        this._offset += 8;
    }

    writeVec(val) {
        this._dv.setFloat64(this._offset, val.x);
        this._offset += 8;
        this._dv.setFloat64(this._offset, val.y);
        this._offset += 8;
    }

    getView() {
        return new DataView(this._dv.buffer, 0, this._offset);
    }
}

class Decoder {
    _dv;
    _offset = 0;
    _buf;

    constructor(buf) {
        this._buf = buf;
        this._dv = new DataView(buf);
    }

    readInt8() {
        let val = this._dv.getInt8(this._offset);
        this._offset += 1;
        return val;
    }

    readUint8() {
        let val = this._dv.getUint8(this._offset);
        this._offset += 1;
        return val;
    }

    readUint16() {
        let val = this._dv.getUint16(this._offset);
        this._offset += 2;
        return val;
    }

    readInt32() {
        let val = this._dv.getInt32(this._offset);
        this._offset += 4;
        return val;
    }

    readFloat64() {
        let val = this._dv.getFloat64(this._offset);
        this._offset += 8;
        return val;
    }

    readVec() {
        let x = this._dv.getFloat64(this._offset);
        this._offset += 8;
        let y = this._dv.getFloat64(this._offset);
        this._offset += 8;
        return new Vec(x, y);
    }

    uint8Array(size) {
        let arr = new Uint8Array(this._buf, this._offset, size);
        this._offset += size;
        return arr;
    }
}

export {Encoder, Decoder};

File: net/encode.go
package net

import (
	"bytes"
	"encoding/binary"

	"github.com/kjander0/ctf/mymath"
)

type Encoder struct {
	Buf    *bytes.Buffer
	Offset int
	Error  error
}

func NewEncoder(buf *bytes.Buffer) Encoder {
	return Encoder{
		Buf: buf,
	}
}

func (e *Encoder) WriteUint8(val uint8) {
	if e.Error != nil {
		return
	}

	e.Error = e.Buf.WriteByte(val)
	e.Offset += 1
}

func (e *Encoder) WriteBytes(val []byte) {
	if e.Error != nil {
		return
	}
	e.Buf.Write(val)
	e.Offset += len(val)
}

func (e *Encoder) WriteUint16(val uint16) {
	if e.Error != nil {
		return
	}

	e.Error = binary.Write(e.Buf, binary.BigEndian, val)
	e.Offset += 2
}

func (e *Encoder) WriteUint16At(val uint16, offset int) {
	if e.Error != nil {
		return
	}

	offsetBuf := bytes.NewBuffer(e.Buf.Bytes()[offset:offset])
	e.Error = binary.Write(offsetBuf, binary.BigEndian, val)
}

func (e *Encoder) WriteFloat64(val float64) {
	if e.Error != nil {
		return
	}
	e.Error = binary.Write(e.Buf, binary.BigEndian, val)
	e.Offset += 8
}

func (e *Encoder) WriteInt32(val int32) {
	if e.Error != nil {
		return
	}
	e.Error = binary.Write(e.Buf, binary.BigEndian, val)
	e.Offset += 4
}

func (e *Encoder) WriteVec(val mymath.Vec) {
	if e.Error != nil {
		return
	}

	e.Error = binary.Write(e.Buf, binary.BigEndian, val.X)
	if e.Error != nil {
		return
	}
	e.Error = binary.Write(e.Buf, binary.BigEndian, val.Y)
	e.Offset += 16
}

type Decoder struct {
	Buf   *bytes.Reader
	Error error
}

func NewDecoder(data []byte) Decoder {
	return Decoder{
		Buf: bytes.NewReader(data),
	}
}

func (d *Decoder) ReadUint8() uint8 {
	if d.Error != nil {
		return 0
	}

	var val uint8
	val, d.Error = d.Buf.ReadByte()
	return val
}

func (d *Decoder) ReadFloat64() float64 {
	if d.Error != nil {
		return 0
	}

	var val float64
	d.Error = binary.Read(d.Buf, binary.BigEndian, &val)
	return val
}

func (d *Decoder) ReadInt32() int32 {
	if d.Error != nil {
		return 0
	}

	var val int32
	d.Error = binary.Read(d.Buf, binary.BigEndian, &val)
	return val
}

func (d *Decoder) ReadVec() mymath.Vec {
	var val mymath.Vec

	if d.Error != nil {
		return val
	}

	d.Error = binary.Read(d.Buf, binary.BigEndian, &val.X)
	if d.Error != nil {
		return val
	}
	d.Error = binary.Read(d.Buf, binary.BigEndian, &val.Y)
	return val
}
